walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,81 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.sql.Time;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    // private static final int SMALLEST_DURATION = 15;
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        /** the request is too long */
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+
+        /** there are no events or no attendes */
+        if (events.isEmpty() || request.getAttendees().isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        } 
+
+        /** Algorithm
+         * 1. Create an array and set of possible meeting times and all events in 30min blocks for the whole day
+         * 2. Remove all the event times from the possible meeting times collection
+         * 3. Piece meeting times together to make longest possible blocks
+         * 4. Remove any blocks which are less than the requested meeting duration
+         * 5. If there are no possible meeting times, repeat without optional attendees (if there are)
+        */
+
+        ArrayList<TimeRange> allTimes = new ArrayList<TimeRange>();
+        for (int i = 0; i < TimeRange.END_OF_DAY; i += 30) {
+            allTimes.add(TimeRange.fromStartDuration(i, 30));
+        }
+
+        Set<TimeRange> eventTimes = new HashSet<TimeRange>();
+        for (Event ev: events) {
+            boolean attendBoth = false;
+            loop: for (String eventAttendee: ev.getAttendees()) {
+                for (String meetingAttendee: request.getAttendees()) {
+                    if (eventAttendee == meetingAttendee) {
```

Comment:
This is tricky.
eventAttendee == meetingAttendee compares the reference of the two string. That is, it returns true only when the two strings points to the exact same string.

To compare the content of the two strings are the same, use eventAttendee.equals(meetingAttendee).

Reply:
Done